mcs/class/Mono.Posix/Test/Mono.Unix.Native/SocketTest.cs

```diff
@@ -0,0 +1,215 @@
+				UnixMarshal.ThrowExceptionForLastError ();
+			try {
+				SetTimeout (sockets[0]);
+				SetTimeout (sockets[1]);
+
+				f (sockets[0], sockets[1]);
+			} finally {
+				if (Syscall.close (sockets[0]) < 0)
+					UnixMarshal.ThrowExceptionForLastError ();
+				if (Syscall.close (sockets[1]) < 0)
+					UnixMarshal.ThrowExceptionForLastError ();
+			}
+		}
+
+		void WithSockets (UnixAddressFamily af, UnixSocketType type, UnixSocketProtocol protocol, Action<int, int> f)
+		{
+			int so1, so2;
+			if ((so1 = Syscall.socket (af, type, protocol)) < 0)
+				UnixMarshal.ThrowExceptionForLastError ();
+			try {
+				if ((so2 = Syscall.socket (af, type, protocol)) < 0)
+					UnixMarshal.ThrowExceptionForLastError ();
+				try {
+					SetTimeout (so1);
+					SetTimeout (so2);
+
+					f (so1, so2);
+				} finally {
+					if (Syscall.close (so2) < 0)
+						UnixMarshal.ThrowExceptionForLastError ();
+				}
+			} finally {
+				if (Syscall.close (so1) < 0)
+					UnixMarshal.ThrowExceptionForLastError ();
+			}
+		}
+
+		[Test]
+		public void Socket ()
+		{
+			int socket;
+			if ((socket = Syscall.socket (UnixAddressFamily.AF_UNIX, UnixSocketType.SOCK_STREAM, 0)) < 0)
+				UnixMarshal.ThrowExceptionForLastError ();
+
+			if (Syscall.close (socket) < 0)
+				UnixMarshal.ThrowExceptionForLastError ();
+		}
+
+		[Test]
+		public void SocketPair ()
+		{
+			var sockets = new int[2];
+			if (Syscall.socketpair (UnixAddressFamily.AF_UNIX, UnixSocketType.SOCK_STREAM, 0, sockets) < 0)
+				UnixMarshal.ThrowExceptionForLastError ();
+
+			if (Syscall.close (sockets[0]) < 0)
+				UnixMarshal.ThrowExceptionForLastError ();
+			if (Syscall.close (sockets[1]) < 0)
+				UnixMarshal.ThrowExceptionForLastError ();
+		}
+
+		[Test]
+		public void SendRecv ()
+		{
+			WithSocketPair ((so1, so2) => {
+				long ret;
+				var buffer1 = new byte[] { 42, 43, 44 };
+				ret = Syscall.send (so1, buffer1, (ulong) buffer1.Length, 0);
+				if (ret < 0)
+					UnixMarshal.ThrowExceptionForLastError ();
+
+				var buffer2 = new byte[1024];
+				ret = Syscall.recv (so2, buffer2, (ulong) buffer2.Length, 0);
+				if (ret < 0)
+					UnixMarshal.ThrowExceptionForLastError ();
+
+				Assert.AreEqual (buffer1.Length, ret);
+				for (int i = 0; i < buffer1.Length; i++)
+					Assert.AreEqual (buffer1[i], buffer2[i]);
+				});
+		}
+
+		[Test]
+		public void SockOpt ()
+		{
+			WithSocketPair ((so1, so2) => {
+				// Set SO_PASSCRED to 1
+				if (Syscall.setsockopt (so1, UnixSocketProtocol.SOL_SOCKET, UnixSocketOptionName.SO_PASSCRED, 1) < 0)
+					UnixMarshal.ThrowExceptionForLastError ();
+
+				// Get and check SO_PASSCRED
+				int value;
+				if (Syscall.getsockopt (so1, UnixSocketProtocol.SOL_SOCKET, UnixSocketOptionName.SO_PASSCRED, out value) < 0)
+					UnixMarshal.ThrowExceptionForLastError ();
+				Assert.AreEqual (value, 1);
+
+				// Set SO_PASSCRED to 0
+				if (Syscall.setsockopt (so1, UnixSocketProtocol.SOL_SOCKET, UnixSocketOptionName.SO_PASSCRED, new byte[10], 4) < 0)
```

Comment:
> Why does new byte[10] set SO_PASSCRED to 0?

Because the length parameter which is passed in is 4, meaning the last 6 bytes of the array are just ignored. The first four bytes are the interpreted as an integer.

> Also, this test will fail on OS X, as it doesn't have SO_PASSCRED (at least, it's not in the man page).

I've switched from SO_PASSCRED to SO_REUSEADDR which is in POSIX. (The only purpose of the code is to test the int and byte[] overload of [sg]etsockopt().)